File: disable_exec.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"regexp"

	"github.com/docker/go-plugins-helpers/authorization"
	"github.com/sirupsen/logrus"
)

type Plugin struct {
	// Will use a Unix socket
	name string
}

func (p *Plugin) AuthZReq(request authorization.Request) authorization.Response {
	response := authorization.Response{
		Allow: true,
		Msg:   "Permission granted",
	}
	if matched, err := regexp.MatchString(`.+/exec$`, request.RequestURI); nil != err {
		logrus.Error("URI check failed.")
	} else {
		if matched {
			if data, err := base64.RawStdEncoding.DecodeString(base64.RawStdEncoding.EncodeToString(request.RequestBody)); nil != err {
				logrus.Error("Request body decoding failed. Error: " + err.Error())
			} else {
				var requestBody map[string]interface{}
				if err = json.Unmarshal(data, &requestBody); nil != err {
					logrus.Error("Loading json request body failed.")
				}
				switch requestBody["User"].(string) {
				case "":
					fallthrough
				case "0":
					fallthrough
				case "root":
					response.Allow = false
					response.Msg = "Permission denied"
					response.Err = "/exec endpoint is currently blocked"
				}
			}
		}
	}
	return response
}

func (p *Plugin) AuthZRes(request authorization.Request) authorization.Response {
	return authorization.Response{
		Allow: true,
	}
}

func main() {
	plugin := &Plugin{name: "authz"}
	handler := authorization.NewHandler(plugin)
	if err := handler.ServeUnix(plugin.name, 0); nil != err {
		logrus.Fatal(err)
	}
}

File: disable_exec.py
from flask import Flask, jsonify, request
import base64
import json
import re
import sys
import os
import logging

logging.basicConfig(format="[%(levelname)s][%(asciitime)s] %(message)s")
logger = logging.getLogger()
logger.setLevel(os.environ.get("LOG_LEVEL", logging.INFO))
PORT = os.environ.get("AUTHZ_PORT", 9090)
plugin = Flask(__name__)
plugin.debug = not not os.environ.get("PLUGIN_DEBUG")

@plugin.route("/Plugin.Activate", methods=["POST"])
def activate():
    return jsonify({"Implements": ["authz"]})

@plugin.route("/AuthZPlugin.AuthZReq", methods=["POST"])
def auth_z_req():
    request_data = json.loads(request.data)
    if re.match(r'.+/exec$', request_data.get("RequestUri")):
        if json.loads(base64.b64decode(request_data.get("RequestBody"))).get("User") in [None, 0, "root"]:
            return jsonify({
                "Allow": False,
                "Msg": "\rExec endpoint is disabled"
            })
    return jsonify({"Allow": True})

@plugin.route("/AuthZPlugin.AuthZRes", methods=["POST"])
def auth_z_res():
    return jsonify({"Allow": True})

plugin.run(port=PORT)